### PR TITLE
Add base64 encoding and decoding to the lua api.

### PR DESCRIPTION
**Pull request by red-001**
see #3804.

### CODE REVIEW

**Comment by ShadowNinja** [comment]
You use newlines inconsistently: one above but not below in lua_api.txt and two more in `l_decode_base64` than in `l_encode_base64`.
Other than that, looks good.

File: src/script/lua_api/l_util.cpp [hunk 1]
```diff
@@ -459,5 +489,8 @@ void ModApiUtil::InitializeAsync(AsyncEngine& engine)
 
 	ASYNC_API_FCT(mkdir);
 	ASYNC_API_FCT(get_dir_list);
+
+	ASYNC_API_FCT(encode_base64);
```
Comment (kwolekr) [inline src/script/lua_api/l_util.cpp]: If this is going to be an async api, it needs to be documented menu_lua_api.txt as well as lua_api.txt.

File: src/script/lua_api/l_util.h [hunk 2]
```diff
@@ -95,6 +95,12 @@ class ModApiUtil : public ModApiBase {
 	// request_insecure_environment()
 	static int l_request_insecure_environment(lua_State *L);
 
+	// encode_base64()
```
Comment (kwolekr) [inline src/script/lua_api/l_util.h]: These two comments above the functions should read [en/de]code_base64(data)

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2420,6 +2420,10 @@ These functions return the leftover itemstack.
     * See documentation on `minetest.compress()` for supported compression methods.
     * currently supported.
     * `...` indicates method-specific arguments. Currently, no methods use this.
+* `minetest.encode_base64(string)`: returns string encoded in base64
+	* Encodes a string in base64.
+* `minetest.decode_base64`: returns string
```
Comment (kwolekr) [inline doc/lua_api.txt]: You're missing the (string) on this line.

File: doc/lua_api.txt [hunk 4]
```diff
@@ -2420,6 +2420,10 @@ These functions return the leftover itemstack.
...
+	* Decodes a string encoded in base64
```
Comment (kwolekr) [inline doc/lua_api.txt]: For indentation in the docs, please use spaces instead.

**Comment by kwolekr** [comment]
Code looks good.  Everything is fixed up and it's just about perfect.

But, I'm wondering if it's a good idea to expose a function from the engine for a task as trivial and indepndent as encoding/decoding a string that could just as easily be done inside of Lua.  In general it's a good idea to keep the surface area of the API as lightweight as possible.
Not to mention, [there are many variants of base64 encoding](https://en.wikipedia.org/wiki/Base64#Variants_summary_table).  If this API catches on then we're going to almost certainly get requests to add options and other bloaty things for alternative encoding types.

At the same time, it seems silly not to leverage functionality that already exists by just adding a small handful of lines so it could be used elsewhere.  I can't decide.

So from me, Code Review +1 but undecided on Concept Approval

**Comment by rubenwardy** [comment]
Would this not be better in builtin? Removes the need for lua -> cpp call.

**Comment by sfan5** [comment]
@rubenwardy 

decoding base64 in pure Lua will surely cause way more overhead than one call from lua -> c++

**Comment by sofar** [comment]
:+1:  on the concept. Mods can use this to safely encode strings into data files, etc.. Very useful.

**Comment by kwolekr** [comment]
Hi red-001, I was wondering what your proposed use case for these new functions would be?

**Comment by OldCoder** [comment]
What might this feature be used for?

**Comment by nerzhul** [comment]
Okay for me :+1:

**Comment by Zeno-** [comment]
Exposing the base64 functions we all ready have to the Lua API? Yeah, seems fine to me as well

**Comment by paramat** [comment]
Zeno- is that an official +1?

Use case in the issue https://github.com/minetest/minetest/issues/3804#issuecomment-221481713

**Comment by Zeno-** [comment]
@paramat I just re-checked that the concerns were addressed and read about the use case, so yes it's an official +1

**Comment by kwolekr** [comment]
Been thinking a bit more about the proposed use case of this for restful web APIs; it makes sense, I approve of this PR.

**Pull request merged**